larpandoracontent/LArControlFlow/BdtBeamParticleIdTool.h

```diff
@@ -0,0 +1,505 @@
+
+namespace lar_content
+{
+
+class AdaBoostDecisionTree;
+
+/**
+ *  @brief  BdtBeamParticleIdTool class
+ */
+class BdtBeamParticleIdTool : public SliceIdBaseTool
+{
+public:
+    /**
+     *  @brief  Constructor
+     */
+    BdtBeamParticleIdTool();
+
+    /**
+     *  @brief  Copy constructor
+     *
+     *  @param  rhs the BdtBeamParticleIdTool to copy
+     */
+    BdtBeamParticleIdTool(const BdtBeamParticleIdTool &rhs);
+
+    /**
+     *  @brief  Assignment operator
+     *
+     *  @param  rhs the BdtBeamParticleIdTool to assign
+     */
+    BdtBeamParticleIdTool &operator=(const BdtBeamParticleIdTool &rhs);
+
+    /**
+     *  @brief  Destructor
+     */
+    ~BdtBeamParticleIdTool();
+
+    void SelectOutputPfos(const pandora::Algorithm *const pAlgorithm, const SliceHypotheses &beamSliceHypotheses, const SliceHypotheses &crSliceHypotheses, pandora::PfoList &selectedPfos);
+
+private:
+    /**
+     *  @brief  Plane class
+     */
+    class Plane
+    {
+    public:
+        /**
+         *  @brief  Constructor, using equation of plane: m_a*x + m_b*y + m_c*z + m_d = 0.
+         *
+         *  @param  normal a Cartesian vector that points in a direction that is normal to the plane
+         *  @param  point a Cartesian vector that corresponds to any point on the plane
+         */
+        Plane(const pandora::CartesianVector &normal, const pandora::CartesianVector &point);
+
+        /**
+         *  @brief  Return the intersection between the plane and a line 
+         *
+         *  @param  a0 point on the line
+         *  @param  a vector pointing along the line
+         */
+        pandora::CartesianVector GetLineIntersection(const pandora::CartesianVector &a0, const pandora::CartesianVector &a) const;
+
+    private:
+        pandora::CartesianVector      m_unitNormal;                         ///< Unit normal to plane
+        pandora::CartesianVector      m_point;                              ///< A point on the plane
+        float                         m_d;                                  ///< Parameter defining a plane
+    };
+
+    typedef std::vector<Plane> PlaneVector;
+
+    /**
+     *  @brief  SliceFeatureParamters class
+     */ 
+    class SliceFeatureParamters
+    {
+    public:
+        /**
+         *  @brief  Constructor
+         */
+        SliceFeatureParamters();
+
+        /**
+         *  @brief  Set TPC geometry information
+         *
+         *  @param  tpcMinX the global TPC volume minimum x extent
+         *  @param  tpcMaxX the global TPC volume maximum x extent
+         *  @param  tpcMinY the global TPC volume minimum y extent
+         *  @param  tpcMaxY the global TPC volume maximum y extent
+         *  @param  tpcMinZ the global TPC volume minimum z extent
+         *  @param  tpcMaxZ the global TPC volume maximum z extent
+         */
+       void SetTPCGeometryInformation(float &tpcMinX, float &tpcMaxX, float &tpcMinY, float &tpcMaxY, float &tpcMinZ, float &tpcMaxZ);
```

Comment:
TPC vs Tpc. Check consistency with other usage.

Arguments should be const float by value, rather than non-const by reference. And all set functions.